common/oauthTokenManager.go

```diff
@@ -300,38 +326,122 @@ func (uotm *UserOAuthTokenManager) UserLogin(tenantID, activeDirectoryEndpoint s
+// If a valid AuthenticationRecord exists, the token is obtained from the azidentity token cache.
+// If the access token has expired, it is refreshed.
+// If the refresh token has expired, the method fails and returns the reason for failure.
+// The fresh token is persisted if either the access token or refresh token has changed.
 func (uotm *UserOAuthTokenManager) getCachedTokenInfo(ctx context.Context) (*OAuthTokenInfo, error) {
-	hasToken, err := uotm.credCache.HasCachedToken()
+
+	// retrieveRecord gets an AuthenticationRecord from the local disk
+	record, err := retrieveRecord()
 	if err != nil {
-		return nil, fmt.Errorf("no cached token found, please log in with azcopy's login command, %v", err)
+		return nil, err
 	}
-	if !hasToken {
-		return nil, errors.New("no cached token found, please log in with azcopy's login command")
+
+	// If the AAuthenticationRecord is empty get a fresh token from our credential cache.
+	if record == (azidentity.AuthenticationRecord{}) {
+		hasToken, err := uotm.credCache.HasCachedToken()
+		if err != nil {
+			return nil, fmt.Errorf("no cached token found, please log in with azcopy's login command, %v", err)
+		}
+		if !hasToken {
+			return nil, errors.New("no cached token found, please log in with azcopy's login command")
+		}
+
+		tokenInfo, err := uotm.credCache.LoadToken()
+		if err != nil {
+			return nil, fmt.Errorf("get cached token failed, %v", err)
+		}
+
+		freshToken, err := tokenInfo.Refresh(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("get cached token failed to ensure token fresh, please log in with azcopy's login command again, %v", err)
+		}
+
+		// Update token cache, if token is updated.
+		if freshToken.Token != tokenInfo.AccessToken.Token {
+			tokenInfo.AccessToken = *freshToken
+			if err := uotm.credCache.SaveToken(*tokenInfo); err != nil {
+				return nil, err
+			}
+		}
+		return tokenInfo, nil
 	}
 
-	tokenInfo, err := uotm.credCache.LoadToken()
+	// Create oAuthTokenInfo from the existing AuthenticationRecord
+	oAuthTokenInfo := &OAuthTokenInfo{
+		Tenant:   record.TenantID,
+		ClientID: record.ClientID,
+	}
+
+	tc, err := oAuthTokenInfo.GetTokenCredential()
 	if err != nil {
-		return nil, fmt.Errorf("get cached token failed, %v", err)
+		return nil, err
 	}
+	scopes := []string{StorageScope}
 
-	freshToken, err := tokenInfo.Refresh(ctx)
+	token, err := tc.GetToken(context.TODO(), policy.TokenRequestOptions{
+		TenantID: record.TenantID,
+		Scopes:   scopes,
+	})
 	if err != nil {
-		return nil, fmt.Errorf("get cached token failed to ensure token fresh, please log in with azcopy's login command again, %v", err)
+		return nil, err
 	}
+	oAuthTokenInfo.AccessToken = token
 
-	// Update token cache, if token is updated.
-	if freshToken.AccessToken != tokenInfo.AccessToken || freshToken.RefreshToken != tokenInfo.RefreshToken {
-		tokenInfo.Token = *freshToken
-		if err := uotm.credCache.SaveToken(*tokenInfo); err != nil {
+	// Update the credential cache with the new token
```

Comment:
Token retreived from AzIdentity is getting stored in credCache below?

Reply:
resolved in the latest commit .